libs/java/auth_core/src/main/java/com/yahoo/athenz/auth/oauth/validator/DefaultOAuthJwtAccessTokenValidator.java

```diff
@@ -0,0 +1,128 @@
+    private Map<String, Set<String>> authorizedClientIds = null;
+
+    /**
+     * create DefaultOAuthJwtAccessTokenValidator
+     * @param  trustedIssuer       trusted issuer (not null, not empty)
+     * @param  requiredAudiences   required audiences (not null, not empty)
+     * @param  requiredScopes      required scopes (not null, not empty)
+     * @param  authorizedClientIds whitelist of authorized client IDs (not null)
+     */
+    public DefaultOAuthJwtAccessTokenValidator(String trustedIssuer, Set<String> requiredAudiences, Set<String> requiredScopes, Map<String, Set<String>> authorizedClientIds) {
+
+        // args checking
+        if (trustedIssuer == null || trustedIssuer.isEmpty()) {
+            throw new IllegalArgumentException("trusted issuers must be configured");
+        }
+        if (requiredAudiences == null || requiredAudiences.isEmpty()) {
+            throw new IllegalArgumentException("required audiences must be configured");
+        }
+        if (requiredScopes == null || requiredScopes.isEmpty()) {
+            throw new IllegalArgumentException("required scopes must be configured");
+        }
+        if (authorizedClientIds == null) {
+            throw new IllegalArgumentException("client ID mapping must be configured");
+        }
+
+        this.trustedIssuer = trustedIssuer;
+        this.requiredAudiences = requiredAudiences;
+        this.requiredScopes = requiredScopes;
+        this.authorizedClientIds = authorizedClientIds;
+    }
+
+    private void verifyIssuer(OAuthJwtAccessToken jwt) throws OAuthJwtAccessTokenException {
+        String issuer = jwt.getIssuer();
+        if (!this.trustedIssuer.equals(issuer)) {
+            // trusted issuer NOT match
+            throw new OAuthJwtAccessTokenException("iss not trusted: got=" + issuer);
+        }
+    }
+    private void verifyAudiences(OAuthJwtAccessToken jwt) throws OAuthJwtAccessTokenException {
+        List<String> audiences = jwt.getAudiences();
+        if (audiences == null || !(new HashSet<>(audiences)).containsAll(this.requiredAudiences)) {
+            // SOME required audiences NOT found
+            String got = (audiences == null) ? "null" : String.join(", ", audiences);
+            throw new OAuthJwtAccessTokenException("required aud not found: got=" + got);
+        }
+    }
+    private void verifyScopes(OAuthJwtAccessToken jwt) throws OAuthJwtAccessTokenException {
+        List<String> scopes = jwt.getScopes();
+        if (scopes == null || !(new HashSet<>(scopes)).containsAll(this.requiredScopes)) {
+            // SOME required scopes NOT found
+            throw new OAuthJwtAccessTokenException("required scope not found: got=" + jwt.getScope());
+        }
+    }
+    private void verifyCertificateThumbprint(OAuthJwtAccessToken jwt, String certificateThumbprint) throws OAuthJwtAccessTokenException {
+        String certThumbprint = jwt.getCertificateThumbprint();
+        if (certificateThumbprint == certThumbprint) {
```

Comment:
String values are compared using `==`, not `equals()`

Reply:
It is to skip checking when both is null.
The actual equal checking it after that.
https://github.com/yahoo/athenz/blob/70d39b4244ae41d5d8ba1d7173be8e3093be7a56/libs/java/auth_core/src/main/java/com/yahoo/athenz/auth/oauth/validator/DefaultOAuthJwtAccessTokenValidator.java#L88-L94

Reply:
I'd suggesting adding a comment explaining that otherwise it does look suspicious

Reply:
added in 9f86ef6d